### PR TITLE
Bloom labs

### PR DESCRIPTION
**Pull request by lragnarsson**

### CODE REVIEW

File: src/Renderer.cpp [hunk 1]
```diff
@@ -375,6 +410,98 @@ void Renderer::ssao_pass()
...
+void Renderer::blur_red_texture(GLuint source_tex, GLuint fbo_tex, GLuint target_fbo, filter_type ft, int iterations)
+{
+    GLuint shader = upload_filter(ft);
+    glUseProgram(shader);
+
+    for (int i=0; i<iterations; i++) {
+        glUniform1i(glGetUniformLocation(shader, "horizontal"), true);
+        filter_pass(source_tex, ping_pong_fbo_red);
+        source_tex = fbo_tex;
+        glUniform1i(glGetUniformLocation(shader, "horizontal"), false);
```
Comment (hansfilipelo) [inline src/Renderer.cpp]: This should be done using two different shaders, one for horizontal and one for vertical. 

As it is right now we get branching in each of them, making calculations for both branches in each. We will improve performance by separating them into two different shaders.

File: src/shaders/blur_rgb_11.frag [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+uniform float[11] kernel;
+uniform float magnitude;
+uniform bool horizontal;
+
+void main()
+{
+    vec2 texel_size = 1.0 / vec2(textureSize(input_tex, 0));
+    vec3 result = kernel[5] * texture(input_tex, TexCoord).rgb;
+
+    if (horizontal) {
```
Comment (hansfilipelo) [inline src/shaders/blur_rgb_11.frag]: See earlier comment about this shader.

File: src/shaders/blur_rgb_5.frag [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+uniform float[5] kernel;
+uniform float magnitude;
+uniform bool horizontal;
+
+void main()
+{
+    vec2 texel_size = 1.0 / vec2(textureSize(input_tex, 0));
+    vec4 result = kernel[2] * texture(input_tex, TexCoord);
+
+    if (horizontal) {
```
Comment (hansfilipelo) [inline src/shaders/blur_rgb_5.frag]: Same applies here.

File: src/shaders/ssao.frag [hunk 4]
```diff
@@ -7,15 +7,12 @@ uniform sampler2D tex_noise;
 
 uniform mat4 view;
 uniform float kernel_radius;
-
-const int MAX_N_SAMPELS = 256;
-
 uniform vec3 samples[256]; // Kernel samples from surrounding screen space geometry
```
Comment (hansfilipelo) [inline src/shaders/ssao.frag]: Should probably be: 

```
uniform vec3 samples[_SSAO_N_SAMPLES_]
```

**Review by hansfilipelo — CHANGES_REQUESTED** [review]

**Review by hansfilipelo — APPROVED** [review]

**Comment by hansfilipelo** [comment]
Post processing kills performance on my MacBook. Post processing stage averages 35-36 ms in 1024x600.

**Comment by hansfilipelo** [comment]
Is it possible to disable bloom?

**Comment by Isakwiberg** [comment]
Post processing averages 55 ms on my MacBook. If I lower the number of blur passes from 3 to 2 in Renderer.cpp line 309 I get 0.2 ms average time! Strange.

**Pull request merged**